<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial for data preparation • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for data preparation">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.7.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial for data preparation</h1>
                        <h4 data-toc-skip class="author">Kaixuan
Luo</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/data_preparation_tutorial.Rmd" class="external-link"><code>vignettes/data_preparation_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>data_preparation_tutorial.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we prepare input data for the analysis.</p>
<p>Load the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="gwas-summary-statistics">GWAS summary statistics<a class="anchor" aria-label="anchor" href="#gwas-summary-statistics"></a>
</h2>
<p>We need a data frame of GWAS summary statistics with the following
columns:</p>
<ul>
<li>chr: chromosome</li>
<li>pos: position<br>
</li>
<li>beta: effect size (if you have Odds Ratio, you will need to
transform it to log(Odds Ratio))<br>
</li>
<li>se: standard error (SE)<br>
</li>
<li>a0: reference allele</li>
<li>a1: association/effect allele</li>
<li>snp: SNP ID (e.g. rsID)</li>
<li>pval: p-value</li>
</ul>
<p>Let’s load a small example GWAS dataset
(<code>test.sumstats.txt.gz</code>), and process the data (see
below).</p>
<p>We can use the <code>vroom</code> package to read the GWAS summary
statistics.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://vroom.r-lib.org" class="external-link">vroom</a></span><span class="op">)</span></span>
<span><span class="va">gwas.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'test.sumstats.txt.gz'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span></span>
<span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://vroom.r-lib.org/reference/vroom.html" class="external-link">vroom</a></span><span class="op">(</span><span class="va">gwas.file</span>, col_names <span class="op">=</span> <span class="cn">TRUE</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">sumstats</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                           var_name                              phase3_1kg_id</span></span>
<span><span class="co">## 1 1_10616_CCGCCGTTGCAAAGGCGCGCCG_C rs376342519:10616:CCGCCGTTGCAAAGGCGCGCCG:C</span></span>
<span><span class="co">## 2                      1_11008_C_G                                1:11008:C:G</span></span>
<span><span class="co">## 3                      1_11012_C_G                                1:11012:C:G</span></span>
<span><span class="co">##   chr position_b37                     a0 a1 bcac_onco2_r2</span></span>
<span><span class="co">## 1   1        10616 CCGCCGTTGCAAAGGCGCGCCG  C      0.837659</span></span>
<span><span class="co">## 2   1        11008                      C  G      0.580657</span></span>
<span><span class="co">## 3   1        11012                      C  G      0.580657</span></span>
<span><span class="co">##   bcac_onco2_eaf_controls bcac_onco2_beta bcac_onco2_se bcac_onco2_P1df_Wald</span></span>
<span><span class="co">## 1                  0.9925         0.00009       0.05784             0.998700</span></span>
<span><span class="co">## 2                  0.0979        -0.04365       0.02002             0.029280</span></span>
<span><span class="co">## 3                  0.0979        -0.04365       0.02002             0.029279</span></span>
<span><span class="co">##   bcac_onco2_P1df_LRT bcac_icogs2_r2       rsIDs</span></span>
<span><span class="co">## 1            0.998700       0.262460 rs376342519</span></span>
<span><span class="co">## 2            0.029368       0.335425           1</span></span>
<span><span class="co">## 3            0.029367       0.335425           1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="ld-blocks">LD blocks<a class="anchor" aria-label="anchor" href="#ld-blocks"></a>
</h2>
<p>For the LD blocks, we need a data frame with four columns:
chromosome, start and end positions, and the indices of the LD
blocks.</p>
<p>We provided the LD blocks from <code>LDetect</code> for the 1000
Genomes (1KG) European population (in hg19).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LD_blocks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'LD.blocks.EUR.hg19.rds'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LD_blocks</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   chr   start     end locus</span></span>
<span><span class="co">## 1   1   10583 1892607     1</span></span>
<span><span class="co">## 2   1 1892607 3582736     2</span></span>
<span><span class="co">## 3   1 3582736 4380811     3</span></span></code></pre>
<p><em>* You can skip this if you only need to run enrichment
analysis.</em></p>
</div>
<div class="section level2">
<h2 id="reference-panel">Reference panel<a class="anchor" aria-label="anchor" href="#reference-panel"></a>
</h2>
<p>Fine-mapping requires linkage-disequilibrium (LD) information.</p>
<p>You could either provide LD matrices or use a reference genotype
panel, which will compute LD matrices internally.</p>
<p>To use the reference genotype panel, we utilize the R package
<code>bigsnpr</code> to read in PLINK files (bed/bim/fam) into R and
match alleles between GWAS summary statistics and reference genotype
panel.</p>
<p>If you have reference genotypes in PLINK format, you can use <a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">bigsnpr::snp_readBed()</a>
to read <code>bed/bim/fam</code> files into a <a href="https://privefl.github.io/bigsnpr/reference/bigSNP-class.html" class="external-link">bigSNP</a>
object and save as a ‘.rds’ file. This ‘.rds’ file could be used for
downstream analyses when attached with .</p>
<p>We provided a <code>bigSNP</code> object of the reference genotype
panel from the 1000 Genomes (1KG) European population. If you are in the
He lab at UChicago, you can load the <code>bigSNP</code> object from RCC
as below.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span>
<span><span class="va">bigSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span>rdsfile <span class="op">=</span> <span class="st">'/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds'</span><span class="op">)</span></span></code></pre></div>
<p><em>* You can skip this if you only need to run enrichment analysis,
or you have LD matrices, or if you only need to run downstream analysis
(e.g. gene mapping) with precomputed fine-mapping result.</em></p>
</div>
<div class="section level2">
<h2 id="process-gwas-summary-statistics">Process GWAS summary statistics<a class="anchor" aria-label="anchor" href="#process-gwas-summary-statistics"></a>
</h2>
<p>We run <code><a href="../reference/process_gwas_sumstats.html">process_gwas_sumstats()</a></code> to process the summary
statistics. This checks and cleans GWAS summary statistics, (optionally)
match alleles in GWAS SNPs with the reference panel from the
<code>bigSNP</code> object, and (optionally) add locus ID from the
<code>LD_Blocks</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_gwas_sumstats.html">process_gwas_sumstats</a></span><span class="op">(</span><span class="va">sumstats</span>, </span>
<span>                                       chr<span class="op">=</span><span class="st">'chr'</span>, </span>
<span>                                       pos<span class="op">=</span><span class="st">'position_b37'</span>, </span>
<span>                                       beta<span class="op">=</span><span class="st">'bcac_onco2_beta'</span>, </span>
<span>                                       se<span class="op">=</span><span class="st">'bcac_onco2_se'</span>,</span>
<span>                                       a0<span class="op">=</span><span class="st">'a0'</span>, </span>
<span>                                       a1<span class="op">=</span><span class="st">'a1'</span>, </span>
<span>                                       snp<span class="op">=</span><span class="st">'phase3_1kg_id'</span>, </span>
<span>                                       pval<span class="op">=</span><span class="st">'bcac_onco2_P1df_Wald'</span>,</span>
<span>                                       LD_Blocks<span class="op">=</span><span class="va">LD_blocks</span>,</span>
<span>                                       bigSNP<span class="op">=</span><span class="va">bigSNP</span><span class="op">)</span></span></code></pre></div>
<p><em>* You do not need to specify <code>bigSNP</code> and
<code>LD_Blocks</code> if you only need to run enrichment
analysis.</em></p>
<p>Check that the summary statistics is processed and has appropriate
columns:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">gwas.sumstats</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   chr   pos a0 a1     beta      se                   snp    pval     zscore</span></span>
<span><span class="co">## 1   1 13110  A  G -0.02066 0.03231           1:13110:G:A 0.52245 -0.6394305</span></span>
<span><span class="co">## 2   1 13116  G  T -0.01255 0.01895 rs201725126:13116:T:G 0.50772 -0.6622691</span></span>
<span><span class="co">## 3   1 13118  G  A -0.01255 0.01895 rs200579949:13118:A:G 0.50772 -0.6622691</span></span>
<span><span class="co">##   locus ss_index bigSNP_index</span></span>
<span><span class="co">## 1     1        3            4</span></span>
<span><span class="co">## 2     1        4            5</span></span>
<span><span class="co">## 3     1        5            6</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
