<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial for enrichment analysis • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for enrichment analysis">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="enrichment_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial for enrichment analysis</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Alan Selewa</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/enrichment_tutorial.Rmd" class="external-link"><code>vignettes/enrichment_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>enrichment_tutorial.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we perform enrichment analysis (using TORUS) using GWAS summary statistics and functional annotations.</p>
<p>Load the packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span></code></pre></div>
<p>Please see the <a href="https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html">data preparation tutorial</a> about preparing GWAS summary statistics. The result from the enrichment analysis will be used in the <a href="https://xinhe-lab.github.io/mapgen/articles/enrichment_finemapping_tutorial.html">fine-mapping tutorial</a>.</p>
<div class="section level2">
<h2 id="enrichment-analysis">Enrichment analysis<a class="anchor" aria-label="anchor" href="#enrichment-analysis"></a>
</h2>
<p>We perform enrichment analysis using the software <code>TORUS</code>.</p>
<p>TORUS needs a SNP annotation file, which contains SNP-level genomic annotations. The annotation file uses a header to specify the number and the nature (categorical or continuous) of the anntations. For example,</p>
<pre><code>SNP   binding_d
chr1.51479  0
chr1.52058  2
chr1.52238  1</code></pre>
<p>The first column with the header “SNP” represents the SNP name. The following columns represent specific annotations. For categorical/discrete annotations, the header should alwasy have a suffix "_d“; whereas for continuous annotations, the header should ends with”_c".</p>
<p>For binary annotations, we could use the function <code><a href="../reference/prepare_torus_input_files.html">prepare_torus_input_files()</a></code> to create a SNP annotation file and a z-score file as TORUS inputs. The function takes the processed GWAS summary statistics obtained from the <a href="https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html">data preparation tutorial</a>, and binary annotation files. The annotation files should be in <code>.bed</code> format, and contain the columns “chr”, “start”, “end”. Chromosomes should be just the numbers (no “chr”).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotation_bed_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'test_bed_dir'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span>,</span>
<span>                                   pattern <span class="op">=</span> <span class="st">'*.bed'</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># annotation_bed_files: Path to the annotation (.bed) files.</span></span>
<span><span class="va">gwas.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'test.processed.sumstats.rds'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">torus.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_torus_input_files.html">prepare_torus_input_files</a></span><span class="op">(</span><span class="va">gwas.sumstats</span>, <span class="va">annotation_bed_files</span>, torus_input_dir <span class="op">=</span> <span class="st">'./torus_input'</span><span class="op">)</span></span>
<span><span class="va">torus_annot_file</span> <span class="op">&lt;-</span> <span class="va">torus.files</span><span class="op">$</span><span class="va">torus_annot_file</span></span>
<span><span class="va">torus_zscore_file</span> <span class="op">&lt;-</span> <span class="va">torus.files</span><span class="op">$</span><span class="va">torus_zscore_file</span></span></code></pre></div>
<p>If you use continuous annotations or annotations with multiple categories, you would need to create your own annotation files.</p>
<p>Now that the appropriate files have been generated, let’s perform the enrichment analysis using <code>TORUS</code>.</p>
<p><code><a href="../reference/run_torus.html">run_torus()</a></code> with <code>option = "est-prior"</code> returns a list with: enrichment estimates (log odds ratio) and 95% confidence intervals for each annotation category, as well as SNP-level priors (which could be used as priors in fine-mapping).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">torus.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_torus.html">run_torus</a></span><span class="op">(</span><span class="va">torus_annot_file</span>, </span>
<span>                       <span class="va">torus_zscore_file</span>,</span>
<span>                       option <span class="op">=</span> <span class="st">'est-prior'</span>,</span>
<span>                       torus_path <span class="op">=</span> <span class="st">'torus'</span><span class="op">)</span> <span class="co"># set the path to 'torus' executable.</span></span>
<span><span class="va">torus.enrich</span> <span class="op">&lt;-</span> <span class="va">torus.res</span><span class="op">$</span><span class="va">enrich</span></span>
<span><span class="va">torus.prior</span> <span class="op">&lt;-</span> <span class="va">torus.res</span><span class="op">$</span><span class="va">snp_prior</span></span></code></pre></div>
<p>If you only want to estimate enrichment but do not need SNP-level priors, you can set <code>option = "est"</code> to save time computing the priors.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">torus.enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_torus.html">run_torus</a></span><span class="op">(</span><span class="va">torus_annot_file</span>, </span>
<span>                          <span class="va">torus_zscore_file</span>,</span>
<span>                          option <span class="op">=</span> <span class="st">'est'</span>,</span>
<span>                          torus_path <span class="op">=</span> <span class="st">'torus'</span><span class="op">)</span><span class="op">$</span><span class="va">enrich</span></span></code></pre></div>
<p>TORUS also gives us the uncertainty of whether each locus contains a causal variant or not. You can set <code>option = "fdr"</code> to get the FDR associated with each locus.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">torus.fdr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_torus.html">run_torus</a></span><span class="op">(</span><span class="va">torus_annot_file</span>, </span>
<span>                       <span class="va">torus_zscore_file</span>,</span>
<span>                       option <span class="op">=</span> <span class="st">'fdr'</span>,</span>
<span>                       torus_path <span class="op">=</span> <span class="st">'torus'</span><span class="op">)</span><span class="op">$</span><span class="va">fdr</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
