<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fine-mapping and LD diagnosis with UK Biobank reference LD matrices • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Fine-mapping and LD diagnosis with UK Biobank reference LD matrices">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="finemapping_ukbb_ld_diagnosis_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/finemapping_ukbb_ld_diagnosis.Rmd" class="external-link"><code>vignettes/finemapping_ukbb_ld_diagnosis.Rmd</code></a></small>
      <div class="hidden name"><code>finemapping_ukbb_ld_diagnosis.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we show an example of performing fine-mapping and LD mismatch diagnosis (using <code>susie_rss</code>) with pre-computed LD matrices from UK Biobank reference.</p>
<p>We have pre-computed the LD matrices of European samples from UK Biobank. They can be downloaded <a href="https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn" class="external-link">here</a>.</p>
<p>Load the packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>Load an example Asthma GWAS summary statistics</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.file</span> <span class="op">&lt;-</span> <span class="st">'/project2/xinhe/shared_data/mapgen/example_data/GWAS/aoa_v3_gwas_ukbsnps_susie_input.txt.gz'</span></span>
<span><span class="va">gwas.sumstats</span> <span class="op">&lt;-</span> <span class="fu">vroom</span><span class="fu">::</span><span class="fu"><a href="https://vroom.r-lib.org/reference/vroom.html" class="external-link">vroom</a></span><span class="op">(</span><span class="va">gwas.file</span>, col_names <span class="op">=</span> <span class="cn">TRUE</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas.sumstats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">##     chr    pos     beta     se a0    a1    snp          pval  zscore locus</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     1 <span style="text-decoration: underline;">693</span>731  0.002<span style="text-decoration: underline;">77</span> 0.015<span style="text-decoration: underline;">6</span> A     G     rs12238997  0.859  0.178      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>     1 <span style="text-decoration: underline;">707</span>522  0.003<span style="text-decoration: underline;">37</span> 0.016<span style="text-decoration: underline;">9</span> G     C     rs371890604 0.841  0.200      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>     1 <span style="text-decoration: underline;">717</span>587 -<span style="color: #BB0000;">0.053</span><span style="color: #BB0000; text-decoration: underline;">8</span>  0.042<span style="text-decoration: underline;">9</span> G     A     rs144155419 0.210 -<span style="color: #BB0000;">1.25</span>       1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>     1 <span style="text-decoration: underline;">723</span>329  0.001<span style="text-decoration: underline;">82</span> 0.128  A     T     rs189787166 0.989  0.014<span style="text-decoration: underline;">3</span>     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span>     1 <span style="text-decoration: underline;">729</span>679  0.005<span style="text-decoration: underline;">77</span> 0.014<span style="text-decoration: underline;">2</span> C     G     rs4951859   0.684  0.407      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span>     1 <span style="text-decoration: underline;">730</span>087 -<span style="color: #BB0000;">0.004</span><span style="color: #BB0000; text-decoration: underline;">65</span> 0.022<span style="text-decoration: underline;">0</span> T     C     rs148120343 0.832 -<span style="color: #BB0000;">0.212</span>      1</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">=</span> <span class="fl">336210</span></span></code></pre></div>
<p>Load LD blocks</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LD_Blocks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'LD.blocks.EUR.hg19.rds'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Obtain a data frame of region_info, rows are LD block coordinates together with the filenames of UKBB reference LD matrices and SNP info.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_UKBB_region_info.html">get_UKBB_region_info</a></span><span class="op">(</span><span class="va">LD_Blocks</span>,</span>
<span>                                    LDREF.dir <span class="op">=</span> <span class="st">"/project2/mstephens/wcrouse/UKB_LDR_0.1_b37"</span>, </span>
<span>                                    prefix <span class="op">=</span> <span class="st">"ukb_b37_0.1"</span><span class="op">)</span></span></code></pre></div>
<p>Read all SNP info from the LD reference</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LD_snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_LD_SNP_info.html">read_LD_SNP_info</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">)</span></span></code></pre></div>
<p>Process GWAS summary statistics and add LD block information, harmonize GWAS summary statistics with LD reference, flip signs when reverse complement matches and remove strand ambiguous variants</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_gwas_sumstats.html">process_gwas_sumstats</a></span><span class="op">(</span><span class="va">gwas.sumstats</span>, </span>
<span>                                       chr<span class="op">=</span><span class="st">'chr'</span>, pos<span class="op">=</span><span class="st">'pos'</span>, beta<span class="op">=</span><span class="st">'beta'</span>, se<span class="op">=</span><span class="st">'se'</span>, </span>
<span>                                       a0<span class="op">=</span><span class="st">'a0'</span>, a1<span class="op">=</span><span class="st">'a1'</span>, snp<span class="op">=</span><span class="st">'snp'</span>, pval<span class="op">=</span><span class="st">'pval'</span>,</span>
<span>                                       LD_snp_info<span class="op">=</span><span class="va">LD_snp_info</span>, </span>
<span>                                       strand_flip<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                                       remove_strand_ambig<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Cleaning summary statistics...</span></span>
<span><span class="co">## Harmonizing GWAS SNPs with LD reference panel...</span></span>
<span><span class="co">## 9732903 variants in sumstats before harmonization...</span></span>
<span><span class="co">## 6419678 variants in both sumstats and LD reference ...</span></span>
<span><span class="co">## Flip signs for 6419677 (100.00%) variants with allele flipped. </span></span>
<span><span class="co">## Remove 994657 (18.33%) strand ambiguous variants. </span></span>
<span><span class="co">## 5425021 variants in sumstats after harmonization...</span></span></code></pre>
<p>Select GWAS significant loci (pval &lt; 5e-8)</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">gwas.sumstats</span><span class="op">$</span><span class="va">pval</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">gwas.sumstats</span> <span class="op">&lt;-</span> <span class="va">gwas.sumstats</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>pval <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">sig.loci</span> <span class="op">&lt;-</span> <span class="va">gwas.sumstats</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">5e-8</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">locus</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sig.loci</span><span class="op">)</span>, <span class="st">"significant loci. \n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 17 significant loci.</span></span></code></pre>
<div class="section level3">
<h3 id="fine-mapping">Fine-mapping<a class="anchor" aria-label="anchor" href="#fine-mapping"></a>
</h3>
<p>Fine-mapping significant loci with pre-computed UKBB LD matrices (we use uniform prior in this tutorial).</p>
<p>If we set <code>save = TRUE</code>, it saves susie result as well as z-scores and R matrix for each locus, which could be used later for diagnosis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected.sumstats</span> <span class="op">&lt;-</span> <span class="va">gwas.sumstats</span><span class="op">[</span><span class="va">gwas.sumstats</span><span class="op">$</span><span class="va">locus</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sig.loci</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">susie.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_finemapping.html">run_finemapping</a></span><span class="op">(</span><span class="va">selected.sumstats</span>, </span>
<span>                                 region_info <span class="op">=</span> <span class="va">region_info</span>, </span>
<span>                                 priortype <span class="op">=</span> <span class="st">'uniform'</span>, n <span class="op">=</span> <span class="va">n</span>, L <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                 save <span class="op">=</span> <span class="cn">TRUE</span>, outputdir <span class="op">=</span> <span class="st">"./test"</span>, outname <span class="op">=</span> <span class="st">"example"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">susie.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_susie_sumstats.html">merge_susie_sumstats</a></span><span class="op">(</span><span class="va">susie.results</span>, <span class="va">sumstats.locus</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/run_finemapping.html">run_finemapping()</a></code> internally runs <code><a href="../reference/susie_finemap_region.html">susie_finemap_region()</a></code> for each of the loci. For finemapping a single locus, we can simply use <code><a href="../reference/susie_finemap_region.html">susie_finemap_region()</a></code> with sumstats and LD information (R and snp_info) for the locus.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">locus</span> <span class="op">&lt;-</span> <span class="va">sig.loci</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">sumstats.locus</span> <span class="op">&lt;-</span> <span class="va">gwas.sumstats</span><span class="op">[</span><span class="va">gwas.sumstats</span><span class="op">$</span><span class="va">locus</span> <span class="op">==</span> <span class="va">locus</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># load LD matrix and SNP info for this locus</span></span>
<span><span class="va">LD_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_UKBB_LDREF.html">load_UKBB_LDREF</a></span><span class="op">(</span><span class="va">LD_Blocks</span>, </span>
<span>                          <span class="va">locus</span>, </span>
<span>                          LDREF.dir <span class="op">=</span> <span class="st">"/project2/mstephens/wcrouse/UKB_LDR_0.1_b37"</span>, </span>
<span>                          prefix <span class="op">=</span> <span class="st">"ukb_b37_0.1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Match GWAS sumstats with LD reference, only keep variants included in LD reference.</span></span>
<span><span class="va">matched.sumstat.LD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/match_gwas_LDREF.html">match_gwas_LDREF</a></span><span class="op">(</span><span class="va">sumstats.locus</span>, <span class="va">LD_ref</span><span class="op">$</span><span class="va">R</span>, <span class="va">LD_ref</span><span class="op">$</span><span class="va">snp_info</span><span class="op">)</span></span>
<span><span class="va">sumstats.locus</span> <span class="op">&lt;-</span> <span class="va">matched.sumstat.LD</span><span class="op">$</span><span class="va">sumstats</span></span>
<span><span class="va">z.locus</span> <span class="op">&lt;-</span> <span class="va">sumstats.locus</span><span class="op">$</span><span class="va">zscore</span></span>
<span><span class="va">R.locus</span> <span class="op">&lt;-</span> <span class="va">matched.sumstat.LD</span><span class="op">$</span><span class="va">R</span></span>
<span><span class="va">snp_info.locus</span> <span class="op">&lt;-</span> <span class="va">matched.sumstat.LD</span><span class="op">$</span><span class="va">snp_info</span></span>
<span></span>
<span><span class="va">susie.locus.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susie_finemap_region.html">susie_finemap_region</a></span><span class="op">(</span><span class="va">sumstats.locus</span>, <span class="va">R.locus</span>, <span class="va">snp_info.locus</span>, n <span class="op">=</span> <span class="va">n</span>, L <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Run susie_rss...</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">susieR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/susieR/man/susie_plots.html" class="external-link">susie_plot</a></span><span class="op">(</span><span class="va">susie.locus.res</span>, y<span class="op">=</span><span class="st">'PIP'</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="finemapping_ukbb_ld_diagnosis_files/figure-html/unnamed-chunk-5-1.png" alt="&amp;nbsp;" width="720"><p class="caption">
 
</p>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">susie.locus.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_susie_sumstats.html">merge_susie_sumstats</a></span><span class="op">(</span><span class="va">susie.locus.res</span>, <span class="va">sumstats.locus</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="ld-mismatch-diagnosis">LD mismatch diagnosis<a class="anchor" aria-label="anchor" href="#ld-mismatch-diagnosis"></a>
</h3>
<p>Inconsistencies between GWAS z-scores and the LD reference could potentially inflate PIPs from SuSiE finemapping when L &gt; 1. It would be helpful to check for potential LD mismatch issues, for example using <a href="https://doi.org/10.1038/s41467-021-27438-7" class="external-link">DENTIST</a> or <a href="https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html" class="external-link">susie_rss</a>.</p>
<p>Here we perform LD mismatch diagnosis for this example locus using <a href="https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html" class="external-link">susie_rss</a>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">condz</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LD_diagnosis_susie_rss.html">LD_diagnosis_susie_rss</a></span><span class="op">(</span><span class="va">z.locus</span>, <span class="va">R.locus</span>, <span class="va">n</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Estimate consistency between the z-scores and reference LD matrix ...</span></span>
<span><span class="co">## Estimated lambda = 7.30936e-05 </span></span>
<span><span class="co">## Compute expected z-scores based on conditional distribution of z-scores ...</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">condz</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="finemapping_ukbb_ld_diagnosis_files/figure-html/unnamed-chunk-6-1.png" alt="&amp;nbsp;" width="720"><p class="caption">
 
</p>
</div>
<p>The estimated lambda is very small, and the LD mismatch diagnosis plot looks good (with no allele flipping detected), which suggests that the GWAS z-scores and reference LD are consistent.</p>
<p>Next, we manually introduces LD mismatches by flipping alleles for 10 variants (with large effects) and see if we can detect those.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seed</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span></span>
<span></span>
<span><span class="va">flip_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sumstats.locus</span><span class="op">$</span><span class="va">zscore</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sumstats.locus.flip</span> <span class="op">&lt;-</span> <span class="va">sumstats.locus</span></span>
<span><span class="va">sumstats.locus.flip</span><span class="op">$</span><span class="va">zscore</span><span class="op">[</span><span class="va">flip_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="va">sumstats.locus</span><span class="op">$</span><span class="va">zscore</span><span class="op">[</span><span class="va">flip_index</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">flip_index</span><span class="op">)</span>, <span class="st">"allele flipped variants:\n"</span>, <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">sumstats.locus.flip</span><span class="op">$</span><span class="va">snp</span><span class="op">[</span><span class="va">flip_index</span><span class="op">]</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 10 allele flipped variants:</span></span>
<span><span class="co">##  rs10184597 rs10189629 rs10203707 rs10207579 rs13392285 rs2110727 rs2215945 rs62154973 rs72823628 rs72823646</span></span></code></pre>
<p>Then, we run fine-mapping including variants with flipped alleles.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">susie.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_finemapping.html">run_finemapping</a></span><span class="op">(</span><span class="va">sumstats.locus.flip</span>, region_info <span class="op">=</span> <span class="va">region_info</span>, priortype <span class="op">=</span> <span class="st">'uniform'</span>, n <span class="op">=</span> <span class="va">n</span>, L <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Finemapping 1 loci...</span></span>
<span><span class="co">## Finemapping locus 193...</span></span>
<span><span class="co">## Run susie_rss...</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">susie.locus.res</span> <span class="op">&lt;-</span> <span class="va">susie.results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">susieR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/susieR/man/susie_plots.html" class="external-link">susie_plot</a></span><span class="op">(</span><span class="va">susie.locus.res</span>, y<span class="op">=</span><span class="st">'PIP'</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="finemapping_ukbb_ld_diagnosis_files/figure-html/unnamed-chunk-8-1.png" alt="&amp;nbsp;" width="720"><p class="caption">
 
</p>
</div>
<p>Let’s compare observed z-scores vs the expected z-scores.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">condz</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LD_diagnosis_susie_rss.html">LD_diagnosis_susie_rss</a></span><span class="op">(</span><span class="va">sumstats.locus.flip</span><span class="op">$</span><span class="va">zscore</span>, R <span class="op">=</span> <span class="va">R.locus</span>, n <span class="op">=</span> <span class="va">n</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Estimate consistency between the z-scores and reference LD matrix ...</span></span>
<span><span class="co">## Estimated lambda = 0.4730198 </span></span>
<span><span class="co">## Compute expected z-scores based on conditional distribution of z-scores ...</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># condz$plot</span></span></code></pre></div>
<p>Detect possible allele flipped variants (logLR &gt; 2 and abs(z) &gt; 2).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">detected_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span><span class="op">$</span><span class="va">logLR</span> <span class="op">&gt;</span> <span class="fl">2</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span><span class="op">$</span><span class="va">z</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Detected %d variants with possible allele flipping"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">detected_index</span><span class="op">)</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Detected 10 variants with possible allele flipping</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Possible allele switched variants:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">sumstats.locus.flip</span><span class="op">$</span><span class="va">snp</span><span class="op">[</span><span class="va">detected_index</span><span class="op">]</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Possible allele switched variants: rs10184597 rs10189629 rs10203707 rs10207579 rs13392285 rs2110727 rs2215945 rs62154973 rs72823628 rs72823646</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span><span class="op">$</span><span class="va">flipped</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span><span class="op">$</span><span class="va">flipped</span><span class="op">[</span><span class="va">flip_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span><span class="op">$</span><span class="va">detected</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span><span class="op">$</span><span class="va">detected</span><span class="op">[</span><span class="va">detected_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%d out of %d flipped variants detected with logLR &gt; 2 and abs(z) &gt; 2. \n"</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">detected_index</span>, <span class="va">flip_index</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">flip_index</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 10 out of 10 flipped variants detected with logLR &gt; 2 and abs(z) &gt; 2.</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span><span class="op">[</span><span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">union</a></span><span class="op">(</span><span class="va">flip_index</span>, <span class="va">detected_index</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              z condmean   condvar z_std_diff    logLR flipped detected</span></span>
<span><span class="co">## 529  -6.965514 6.321319 0.4847394 -19.083908 8.381969       1        1</span></span>
<span><span class="co">## 1065 -2.567840 2.471008 0.4833665  -7.247579 8.078109       1        1</span></span>
<span><span class="co">## 836  -2.532926 2.380152 0.4935893  -6.993116 8.042908       1        1</span></span>
<span><span class="co">## 1048 -2.611843 2.493572 0.4837408  -7.340484 8.074479       1        1</span></span>
<span><span class="co">## 218  -3.510579 3.281118 0.5026859  -9.579216 8.133171       1        1</span></span>
<span><span class="co">## 115  -4.280961 4.094475 0.4988993 -11.857713 8.320665       1        1</span></span>
<span><span class="co">## 1365 -3.318277 3.265362 0.5069206  -9.246898 8.196465       1        1</span></span>
<span><span class="co">## 433  -6.686770 6.123010 0.4997885 -18.119598 8.451238       1        1</span></span>
<span><span class="co">## 624  -6.934847 6.350230 0.4841194 -19.093601 8.509518       1        1</span></span>
<span><span class="co">## 142  -4.512544 4.337933 0.4937149 -12.595882 8.385750       1        1</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">condmean</span>, y <span class="op">=</span> <span class="va">z</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">flipped</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span> <span class="op">=</span> <span class="st">"black"</span>, <span class="st">"1"</span> <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Expected value"</span>, y <span class="op">=</span> <span class="st">"Observed z scores"</span>, color <span class="op">=</span> <span class="st">"Simulated allele flipping"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="finemapping_ukbb_ld_diagnosis_files/figure-html/diagnosis-plots-1.png" alt="&amp;nbsp;" width="720"><p class="caption">
 
</p>
</div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">condz</span><span class="op">$</span><span class="va">conditional_dist</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">condmean</span>, y <span class="op">=</span> <span class="va">z</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">detected</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span> <span class="op">=</span> <span class="st">"black"</span>, <span class="st">"1"</span> <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Expected value"</span>, y <span class="op">=</span> <span class="st">"Observed z scores"</span>, color <span class="op">=</span> <span class="st">"Detected allele flipping"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="finemapping_ukbb_ld_diagnosis_files/figure-html/diagnosis-plots-2.png" alt="&amp;nbsp;" width="720"><p class="caption">
 
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
