<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial for fine-mapping with functional priors • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for fine-mapping with functional priors">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="finemapping_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial for fine-mapping with functional priors</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Alan Selewa</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/finemapping_tutorial.Rmd" class="external-link"><code>vignettes/finemapping_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>finemapping_tutorial.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we perform fine-mapping (using SuSiE) with functional priors computed from the enrichment analysis using <code>TORUS</code>.</p>
<p>Load the packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>Please see the <a href="https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html">data preparation tutorial</a> about preparing GWAS summary statistics.</p>
<p>We need a <code>bigSNP</code> object for the reference genotype panel, which will be used for computing LD matrices. See the <a href="https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html">data preparation tutorial</a> for details.</p>
</div>
<div class="section level2">
<h2 id="snp-level-priors">SNP-level priors<a class="anchor" aria-label="anchor" href="#snp-level-priors"></a>
</h2>
<p>Let’s load the processed GWAS summary statistics obtained from the <a href="https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html">data preparation tutorial</a> and the SNP-level priors using <code><a href="../reference/run_torus.html">run_torus()</a></code> obtained from the <a href="https://xinhe-lab.github.io/mapgen/articles/enrichment_finemapping_tutorial.html">enrichment analysis tutorial</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gwas.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'test.processed.sumstats.rds'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">torus.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'test.torus.res.rds'</span>, package<span class="op">=</span><span class="st">'mapgen'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can add the SNP-level priors (column called ‘torus_prior’) to our summary statistics.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumstats.for.susie</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_susie_data_with_torus_result.html">prepare_susie_data_with_torus_result</a></span><span class="op">(</span>sumstats <span class="op">=</span> <span class="va">gwas.sumstats</span>, </span>
<span>                                                           torus_prior <span class="op">=</span> <span class="va">torus.res</span><span class="op">$</span><span class="va">snp_prior</span><span class="op">)</span></span></code></pre></div>
<p>We could set <code>filter == "pval"</code> to limit loci with GWAS p-value cutoff (default: 5e-8), or set <code>filter == "torus_fdr"</code> by the FDR cutoff (default: 0.1) from running <code><a href="../reference/run_torus.html">run_torus()</a></code> with <code>option = "fdr"</code>.</p>
</div>
<div class="section level2">
<h2 id="fine-mapping">Fine-mapping<a class="anchor" aria-label="anchor" href="#fine-mapping"></a>
</h2>
<p>We can perform fine-mapping with SNP-level priors using <code>SuSiE</code>, you will need the <code>susieR</code> package for the steps below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/susieR" class="external-link">susieR</a></span><span class="op">)</span></span></code></pre></div>
<p>To use SNP-level priors computed from <code>TORUS</code>, we use <code><a href="../reference/run_finemapping.html">run_finemapping()</a></code> with <code>priortype = 'torus'</code>.</p>
<p>You can also run fine-mapping using uniform prior (without SNP-level priors) with <code>priortype = 'uniform'</code>. <em>(You don’t need to run <code>TORUS</code> if you run fine-mapping using uniform prior.)</em></p>
<p>You can use the bigSNP object prepared earlier in the <a href="https://xinhe-lab.github.io/mapgen/articles/data_preparation_tutorial.html">data preparation tutorial</a>. If you provide bigSNP instead of LD matrices, it will compute R matrices using the genotype data in the bigSNP object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># you will need the bigsnpr package if you use the bigSNP object.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span>
<span><span class="va">susie.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_finemapping.html">run_finemapping</a></span><span class="op">(</span>sumstats <span class="op">=</span> <span class="va">sumstats.for.susie</span>, </span>
<span>                             bigSNP <span class="op">=</span> <span class="va">bigSNP</span>, </span>
<span>                             priortype <span class="op">=</span> <span class="st">'torus'</span>, </span>
<span>                             L <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>We did not provide sample size (n) for this toy example, but it is highly encouraged to provide that (set argument <code>n</code> = sample size) in the real data analysis.</p>
<p>If you have the LD matrices, you can use those instead of the bigSNP object.</p>
<p>We have pre-computed the LD matrices of European samples from UK Biobank. They can be downloaded <a href="https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn" class="external-link">here</a>.</p>
<p>See <a href="https://xinhe-lab.github.io/mapgen/articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping with UK Biobank LD tutorial</a> and [LD diagnosis tutorial][ld-diagnosis-tutorial] for details.</p>
<p>We ran <code>SuSiE</code> with <code>L = 1</code> here, which allows a single causal signal for each LD block and is robust to mismatching LD patterns.</p>
<p><code>susie.res</code> is a list of <code>SuSiE</code> results, one for each locus. We can merge GWAS summary statistics with SuSiE result:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap.sumstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_susie_sumstats.html">merge_susie_sumstats</a></span><span class="op">(</span><span class="va">susie.res</span>, <span class="va">sumstats.for.susie</span><span class="op">)</span></span></code></pre></div>
<p>We now have a new column called <code>susie_pip</code>, which is the probability of a SNP being causal estimated using <code>SuSiE</code>.</p>
<p><em>Note</em>: Inconsistencies between GWAS z-scores and the LD reference could potentially inflate PIPs from SuSiE finemapping. It would be helpful to check potential LD mismatch issue, for example using <a href="https://doi.org/10.1038/s41467-021-27438-7" class="external-link">DENTIST</a> or the <a href="https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html" class="external-link">diagnostic</a> from <code>susie_rss</code>.</p>
<p>See <a href="https://xinhe-lab.github.io/mapgen/articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping with UK Biobank LD tutorial</a> for details.</p>
<p>So to be conservative, you could use <code>L = 1</code> if you have “out-of-sample” LD reference, and focus on the loci with significant GWAS signals (e.g. <code>pval &lt; 5e-8</code>).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
