<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene mapping tutorial • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene mapping tutorial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="gene_mapping_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene mapping tutorial</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/gene_mapping_tutorial.Rmd" class="external-link"><code>vignettes/gene_mapping_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>gene_mapping_tutorial.Rmd</code></div>

    </div>

    
    
<p>Despite our fine-mapping efforts, there remained considerable uncertainty of causal variants in most loci. Even if the causal variants are known, assigning target genes can be difficult due to long-range regulation of enhancers.</p>
<p>Our gene mapping procedure prioritizes target genes:</p>
<ol style="list-style-type: decimal">
<li><p>For every putative causal SNP, we assign a weight to each nearby gene, considering multiple ways the SNP may affect a gene. The weight of a gene can be viewed as the probability that the SNP affects that gene.</p></li>
<li><p>For each gene, we then aggregate the causal evidence of all SNPs likely targeting this gene, expressed as the weighted sum of the PIPs of all these SNPs. To ensure that the causal evidence of a variant is not counted multiple times when it targets multiple genes, we normalize the SNP-to-gene weights in this calculation. The resulting “gene PIP” approximates the probability of a gene being causal. Similar to variant-level fine-mapping, we also define a “credible gene set”, the set of genes that capture the causal signal at a locus with high probability.</p></li>
</ol>
<p>The weights of SNP-gene pairs reflect the strength of biological evidence linking SNPs to genes. For a SNP in an exon/promoter or in a regulatory region linked to a particular gene (“enhancer loops”), we assign a weight of 1 to that gene.</p>
<p>When a SNP cannot be linked to any gene in these ways, its target genes are assigned using a distance weighted function so that nearby genes receive higher weights.</p>
<p>See the Methods section in our paper for details.</p>
<div class="figure">
<img src="../reference/figures/gene.mapping.diagram.png" alt="Schematic of gene-level PIP calculation. s: SNP, g: gene" width="75%"><p class="caption">
Schematic of gene-level PIP calculation. s: SNP, g: gene
</p>
</div>
<p>In this tutorial, we show the gene mapping procedure using the fine-mapping result from our AFib study.</p>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>Our gene mapping procedure takes the following input data:</p>
<ul>
<li>Fine-mapping summary statistics.</li>
<li>Gene annotations, exons, promoters, introns, UTRs, etc.</li>
<li>Functional annotation data, including: ABC scores, PC-HiC loops, open chromatin regions (OCRs), etc.</li>
</ul>
<p>Load R packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="fine-mapping-summary-statistics">Fine-mapping summary statistics<a class="anchor" aria-label="anchor" href="#fine-mapping-summary-statistics"></a>
</h3>
<p>Here we use fine-mapping summary statistics from our AFib study.</p>
<p>Keep SNPs with PIP &gt; 1e-5, rename columns, and save as a GRanges object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"AF.finemapping.sumstats.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_finemapping_sumstats.html">process_finemapping_sumstats</a></span><span class="op">(</span><span class="va">finemapstats</span>, </span>
<span>                                             snp <span class="op">=</span> <span class="st">'snp'</span>, </span>
<span>                                             chr <span class="op">=</span> <span class="st">'chr'</span>, </span>
<span>                                             pos <span class="op">=</span> <span class="st">'pos'</span>, </span>
<span>                                             pip <span class="op">=</span> <span class="st">'susie_pip'</span>, </span>
<span>                                             pval <span class="op">=</span> <span class="st">'pval'</span>, </span>
<span>                                             zscore <span class="op">=</span> <span class="st">'zscore'</span>, </span>
<span>                                             cs <span class="op">=</span> <span class="st">'cs'</span>, </span>
<span>                                             locus <span class="op">=</span> <span class="st">'locus'</span>,  </span>
<span>                                             pip.thresh <span class="op">=</span> <span class="fl">1e-5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Processing fine-mapping summary statistics ...</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'utils::download.file' by</span></span>
<span><span class="co">## 'restfulr::download.file' when loading 'rtracklayer'</span></span></code></pre>
<pre><code><span><span class="co">## Select SNPs with PIP &gt; 1e-05</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For simplicity, only keep the following columns in the steps below</span></span>
<span><span class="va">cols.to.keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'snp'</span>,<span class="st">'chr'</span>,<span class="st">'pos'</span>, <span class="st">'pip'</span>, <span class="st">'pval'</span>, <span class="st">'zscore'</span>, <span class="st">'cs'</span>, <span class="st">'locus'</span><span class="op">)</span></span>
<span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="va">finemapstats</span><span class="op">[</span>, <span class="va">cols.to.keep</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finemapstats</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 8 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |         snp       chr       pos         pip</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1  10161265      * | rs189262553         1  10161265 0.005541892</span></span>
<span><span class="co">##   [2]     chr1  10167425      * | rs187585530         1  10167425 0.031227600</span></span>
<span><span class="co">##   [3]     chr1  10727192      * | rs191213578         1  10727192 0.000159041</span></span>
<span><span class="co">##            pval    zscore        cs     locus</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]   7.89245  -5.69091         0         7</span></span>
<span><span class="co">##   [2]   8.11250  -5.77273         0         7</span></span>
<span><span class="co">##   [3]   4.53313  -4.17653         0         7</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 20 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="genomic-annotations">Genomic annotations<a class="anchor" aria-label="anchor" href="#genomic-annotations"></a>
</h3>
<p>To link SNPs to genes, we use a hierarchical approach to assign the following genomic annotation categories:</p>
<ol style="list-style-type: decimal">
<li>Exons/(active) promoters category</li>
<li>Enhancer loops category (ABC, PC-HiC, etc.)</li>
<li>Enhancer regions category (open chromatin regions, etc.)</li>
<li>Introns/UTRs category</li>
</ol>
<div class="section level4">
<h4 id="gene-annotations">Gene annotations<a class="anchor" aria-label="anchor" href="#gene-annotations"></a>
</h4>
<p>You can download gene annotations (GTF file) from <a href="https://www.gencodegenes.org" class="external-link">GENCODE</a>, then make gene annotations using the GTF file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="st">'/project2/xinhe/shared_data/gencode/gencode.v19.annotation.gtf.gz'</span></span>
<span><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_genomic_annots.html">make_genomic_annots</a></span><span class="op">(</span><span class="va">gtf_file</span><span class="op">)</span></span></code></pre></div>
<p>We included gene annotations (hg19) in the package, downloaded from <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz" class="external-link">GENCODE release 19</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genomic.annots.hg19.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene.annots</span> <span class="op">&lt;-</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">genes</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="open-chromatin-regions-ocrs">Open chromatin regions (OCRs)<a class="anchor" aria-label="anchor" href="#open-chromatin-regions-ocrs"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OCRs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"OCRs.hg19.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="exonspromoters-category">Exons/promoters category<a class="anchor" aria-label="anchor" href="#exonspromoters-category"></a>
</h4>
<p>In this example, we include exons and active promoters in this category for gene mapping. We defined active promoters as promoters overlapping with OCRs (overlap at least 100 bp).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">active_promoters</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">promoters</span>, </span>
<span>                                              <span class="va">OCRs</span>, </span>
<span>                                              minoverlap <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">exons_promoters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>exons <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">exons</span>, </span>
<span>                                       active_promoters <span class="op">=</span> <span class="va">active_promoters</span><span class="op">)</span></span></code></pre></div>
<p>If you don’t have OCR data, you can simply use exons and promoters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">exons_promoters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>exons <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">exons</span>, </span>
<span>                                       promoters <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">promoters</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="enhancer-loops-category">Enhancer loops category<a class="anchor" aria-label="anchor" href="#enhancer-loops-category"></a>
</h4>
<p>We need chromatin loop data, such as PC-HiC or ABC scores, with the following columns, and save as a GRanges object.</p>
<ul>
<li>chr: chromosome of the regulatory element</li>
<li>start: start position of the regulatory element</li>
<li>end: end position of the regulatory element</li>
<li>promoter_start: start position of the gene promoter or TSS position</li>
<li>promoter_end: end position of the gene promoter or TSS position</li>
<li>gene_name: target gene name</li>
<li>score: numeric interaction score (optional)</li>
</ul>
<p><strong>Promoter-capture HiC (PC-HiC)</strong></p>
<p>Here we use promoter-capture HiC (PC-HiC) data from cardiomyocytes (CMs). You may skip this if you do not have relevant PC-HiC data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcHiC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"pcHiC.CM.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pcHiC</span> <span class="op">&lt;-</span> <span class="va">pcHiC</span><span class="op">[</span><span class="va">pcHiC</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gene.annots</span><span class="op">$</span><span class="va">gene_name</span>, <span class="op">]</span> <span class="co"># restrict to protein coding genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pcHiC</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand | promoter_start promoter_end</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1 10056320-10057012      * |       10002213     10004089</span></span>
<span><span class="co">##   [2]     chr1 10062800-10063261      * |       10002213     10004089</span></span>
<span><span class="co">##   [3]     chr1 10085205-10085681      * |       10002213     10004089</span></span>
<span><span class="co">##         gene_name     score</span></span>
<span><span class="co">##       &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]      NMNAT1      6.33</span></span>
<span><span class="co">##   [2]      NMNAT1      7.37</span></span>
<span><span class="co">##   [3]      NMNAT1      7.28</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p><strong>ABC scores</strong></p>
<p>The <code><a href="../reference/process_ABC.html">process_ABC()</a></code> function processes a data frame of ABC scores from Nasser <em>et al.</em> <em>Nature</em> 2021, and converts it to a GRanges object.</p>
<p>Here we use ABC scores from heart ventricle (from Nasser <em>et al.</em> <em>Nature</em> 2021). You may skip this if you do not have relevant ABC scores.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/data.table/man/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"heart_ventricle-ENCODE_ABC.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_ABC.html">process_ABC</a></span><span class="op">(</span><span class="va">ABC</span>, full.element <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="va">ABC</span><span class="op">[</span><span class="va">ABC</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gene.annots</span><span class="op">$</span><span class="va">gene_name</span>, <span class="op">]</span> <span class="co"># restrict to protein coding genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ABC</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames        ranges strand | promoter_start promoter_end   gene_name</span></span>
<span><span class="co">##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;integer&gt;    &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 888243-888743      * |         894679       894679       NOC2L</span></span>
<span><span class="co">##   [2]     chr1 908361-908861      * |         895966       895966      KLHL17</span></span>
<span><span class="co">##   [3]     chr1 908361-908861      * |         901876       901876     PLEKHN1</span></span>
<span><span class="co">##           score</span></span>
<span><span class="co">##       &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]  0.015224</span></span>
<span><span class="co">##   [2]  0.017673</span></span>
<span><span class="co">##   [3]  0.041100</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="enhancer-regions">Enhancer regions<a class="anchor" aria-label="anchor" href="#enhancer-regions"></a>
</h4>
<p>Here we define enhancer regions by OCRs. You can also use histone mark data to define enhancer regions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">enhancer_regions</span> <span class="op">&lt;-</span> <span class="va">OCRs</span><span class="op">[</span><span class="va">OCRs</span><span class="op">$</span><span class="va">peakType</span><span class="op">!=</span><span class="st">"Promoter"</span>,<span class="op">]</span></span></code></pre></div>
<p>Considering the fact that PC-HiC and HiC loops may miss contacts between close regions due to technical reasons, here we also consider enhancer regions within 20 kb of active promoters as “nearby interactions”.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nearby20kb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nearby_interactions.html">nearby_interactions</a></span><span class="op">(</span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">enhancer_regions</span>,</span>
<span>                                  <span class="va">active_promoters</span>, </span>
<span>                                  max.dist <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span></span></code></pre></div>
<p>We include ABC, PC-HiC and nearby interactions in the “enhancer loop” category.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span><span class="op">$</span><span class="va">enhancer_loops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ABC <span class="op">=</span> <span class="va">ABC</span>, pcHiC <span class="op">=</span> <span class="va">pcHiC</span>, nearby20kb <span class="op">=</span> <span class="va">nearby20kb</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">genomic.annots</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  Length Class   Mode</span></span>
<span><span class="co">## exons            195913 GRanges S4  </span></span>
<span><span class="co">## introns          189762 GRanges S4  </span></span>
<span><span class="co">## UTRs              50889 GRanges S4  </span></span>
<span><span class="co">## promoters         19430 GRanges S4  </span></span>
<span><span class="co">## genes             19430 GRanges S4  </span></span>
<span><span class="co">## exons_promoters       2 -none-  list</span></span>
<span><span class="co">## enhancer_regions 327014 GRanges S4  </span></span>
<span><span class="co">## enhancer_loops        3 -none-  list</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="run-gene-mapping">Run gene mapping<a class="anchor" aria-label="anchor" href="#run-gene-mapping"></a>
</h3>
<p>Parameters:</p>
<ul>
<li>finemapstats: A GRanges object of processed fine-mapping result.</li>
<li>genomic.annots: A list of GRanges objects of different annotation categories.</li>
<li>intron.mode: If <code>TRUE</code>, assign intronic SNPs to genes containing the introns.</li>
<li>d0: distance weight parameter (default: 50000).</li>
<li>exon.weight: weight for exons/promoters (default = 1).</li>
<li>loop.weight: weight for enhancer loops (default = 1).</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene.mapping.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_gene_pip.html">compute_gene_pip</a></span><span class="op">(</span><span class="va">finemapstats</span>, </span>
<span>                                     <span class="va">genomic.annots</span>, </span>
<span>                                     intron.mode <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                     d0 <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                                     exon.weight <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                     loop.weight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Map SNPs to genes and assign weights ...</span></span>
<span><span class="co">## Assign SNPs in exons and promoters ...</span></span>
<span><span class="co">## Assign SNPs to linked genes through enhancer loops ...</span></span>
<span><span class="co">## Enhancer loops include: ABC, pcHiC, nearby20kb </span></span>
<span><span class="co">## Assign SNPs in enhancer regions to genes by distance weighting ...</span></span>
<span><span class="co">## Assign SNPs in UTRs (excluding enhancer regions) to the UTR genes ... </span></span>
<span><span class="co">## Assign the rest of SNPs (intergenic) to genes by distance weighting ...</span></span>
<span><span class="co">## Compute gene PIP ... </span></span>
<span><span class="co">## Done.</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.mapping.res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          snp chr       pos          pip    pval    zscore cs locus gene_name</span></span>
<span><span class="co">## 1  rs6658920   1  22257739 3.991827e-03 6.40472  5.057971  0    15     HSPG2</span></span>
<span><span class="co">## 2  rs6658920   1  22257739 3.991827e-03 6.40472  5.057971  0    15     USP48</span></span>
<span><span class="co">## 3   rs213734   1  41575699 4.560662e-05 3.56067 -3.611940  0    25     SCMH1</span></span>
<span><span class="co">## 4  rs2793374   1 205647508 6.666840e-05 4.70421  4.268657  0   106   SLC45A3</span></span>
<span><span class="co">## 5 rs12047038   1 205718468 4.506310e-04 4.26074 -4.040541  0   106    NUCKS1</span></span>
<span><span class="co">## 6  rs7576657   2  60779478 1.328678e-05 3.19620 -3.432432  0   173    BCL11A</span></span>
<span><span class="co">##         category weight  frac_pip    gene_pip</span></span>
<span><span class="co">## 1 ABC,nearby20kb      1 0.3333333 0.494801999</span></span>
<span><span class="co">## 2            ABC      1 0.3333333 0.001727402</span></span>
<span><span class="co">## 3            ABC      1 1.0000000 0.075123247</span></span>
<span><span class="co">## 4            ABC      1 1.0000000 0.017269657</span></span>
<span><span class="co">## 5 ABC,nearby20kb      1 1.0000000 0.554064134</span></span>
<span><span class="co">## 6 ABC,nearby20kb      1 1.0000000 0.000439048</span></span></code></pre>
<p>Gene-level result table</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene.pip.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_gene_level_result.html">extract_gene_level_result</a></span><span class="op">(</span><span class="va">gene.mapping.res</span>, <span class="va">gene.annots</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.pip.res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    chr       pos gene_name    gene_pip</span></span>
<span><span class="co">## 1 chr1  22263790     HSPG2 0.494801999</span></span>
<span><span class="co">## 2 chr1  22110099     USP48 0.001727402</span></span>
<span><span class="co">## 3 chr1  41707826     SCMH1 0.075123247</span></span>
<span><span class="co">## 4 chr1 205649587   SLC45A3 0.017269657</span></span>
<span><span class="co">## 5 chr1 205719404    NUCKS1 0.554064134</span></span>
<span><span class="co">## 6 chr2  60780702    BCL11A 0.000439048</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%d genes with PIP &gt;= 0.8"</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gene.pip.res</span><span class="op">$</span><span class="va">gene_name</span><span class="op">[</span><span class="va">gene.pip.res</span><span class="op">$</span><span class="va">gene_pip</span> <span class="op">&gt;=</span> <span class="fl">0.8</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 46 genes with PIP &gt;= 0.8</span></span></code></pre>
<p>80% credible gene sets</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene.cs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_cs.html">gene_cs</a></span><span class="op">(</span><span class="va">gene.mapping.res</span>, by.locus <span class="op">=</span> <span class="cn">TRUE</span>, gene.cs.coverage <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene.cs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   locus      gene_cs          gene_cs_locus_pip top_gene top_locus_gene_pip</span></span>
<span><span class="co">## 1     7        CASZ1               CASZ1(0.882)    CASZ1              0.882</span></span>
<span><span class="co">## 2    15 HSPG2,CELA3B HSPG2(0.495),CELA3B(0.486)    HSPG2              0.495</span></span>
<span><span class="co">## 3    25 SLFNL1,CTPS1 SLFNL1(0.625),CTPS1(0.259)   SLFNL1              0.625</span></span>
<span><span class="co">## 4    31        BEND5               BEND5(0.997)    BEND5              0.997</span></span>
<span><span class="co">## 5    32  FAF1,CDKN2C   FAF1(0.41),CDKN2C(0.405)     FAF1              0.410</span></span>
<span><span class="co">## 6    33 TTC39A,EPS15  TTC39A(0.447),EPS15(0.38)   TTC39A              0.447</span></span>
<span><span class="co">##   top_gene_pip</span></span>
<span><span class="co">## 1        0.882</span></span>
<span><span class="co">## 2        0.495</span></span>
<span><span class="co">## 3        0.625</span></span>
<span><span class="co">## 4        0.997</span></span>
<span><span class="co">## 5        0.426</span></span>
<span><span class="co">## 6        0.448</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="gene-manhattan-plot">Gene Manhattan plot<a class="anchor" aria-label="anchor" href="#gene-manhattan-plot"></a>
</h3>
<p>label genes with gene PIP &gt; 0.8.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene_manhattan_plot.html">gene_manhattan_plot</a></span><span class="op">(</span><span class="va">gene.pip.res</span>, gene.pip.thresh <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p><img src="gene_mapping_tutorial_files/figure-html/gene-manhattan-plot-1.png" width="768"></p>
<p><em>Note</em>: in some cases where a gene spans two nearby LD blocks and/or is linked to SNPs in multiple credible sets (L &gt; 1), the gene PIP may exceed 1, which can be interpreted as the expected number of causal variants targeting the gene (see Methods of our paper for details).</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
