<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Partition fine-mapping PIPs by annotation categories â€¢ mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Partition fine-mapping PIPs by annotation categories">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="partition_pip_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Partition fine-mapping PIPs by annotation categories</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/partition_pip_tutorial.Rmd" class="external-link"><code>vignettes/partition_pip_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>partition_pip_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="partitioning-pips-by-cell-types-and-annotation-categories">Partitioning PIPs by cell types and annotation categories<a class="anchor" aria-label="anchor" href="#partitioning-pips-by-cell-types-and-annotation-categories"></a>
</h2>
<p>Here, we show an example using AFib GWAS finemapping result and scATAC-seq data from our heart study. We assigned the likely cell type(s) through which the causal variants act in each locus using fine-mapped SNPs, and cell-type specific open chromatin regions (OCRs) obtained from scATAC-seq data.</p>
<p><strong>Required input data</strong>:</p>
<ul>
<li>Fine-mapped summary statistics.</li>
<li>A list of GRanges objects of annotation regions.</li>
</ul>
<p>Load R packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span></code></pre></div>
<p>Load fine-mapping summary statistics from the AFib study Keep SNPs with PIP &gt; 1e-5, rename columns, and save as a GRanges object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"AF.finemapping.sumstats.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_finemapping_sumstats.html">process_finemapping_sumstats</a></span><span class="op">(</span><span class="va">finemapstats</span>, </span>
<span>                                             snp <span class="op">=</span> <span class="st">'snp'</span>, chr <span class="op">=</span> <span class="st">'chr'</span>, </span>
<span>                                             pos <span class="op">=</span> <span class="st">'pos'</span>, pip <span class="op">=</span> <span class="st">'susie_pip'</span>, </span>
<span>                                             pval <span class="op">=</span> <span class="st">'pval'</span>, zscore <span class="op">=</span> <span class="st">'zscore'</span>, </span>
<span>                                             cs <span class="op">=</span> <span class="st">'cs'</span>, locus <span class="op">=</span> <span class="st">'locus'</span>,  </span>
<span>                                             pip.thresh <span class="op">=</span> <span class="fl">1e-5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co"># Processing fine-mapping summary statistics ...</span></span></code></pre>
<pre><code><span><span class="co"># Warning: replacing previous import 'utils::download.file' by</span></span>
<span><span class="co"># 'restfulr::download.file' when loading 'rtracklayer'</span></span></code></pre>
<pre><code><span><span class="co"># Select SNPs with PIP &gt; 1e-05</span></span></code></pre>
<p>We can partition PIPs into different functional annotation categories.</p>
<p>Load genomic annotations (hg19).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genomic.annots.hg19.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Load cell-type specific open chromatin regions (OCRs) (hg19).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OCRs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"OCRs.hg19.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Create a list of the annotations, with priority in the order of OCRs, UTRS, Exons, and Introns.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annots.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>OCRs <span class="op">=</span> <span class="va">OCRs</span>,</span>
<span>                    UTRs <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">UTRs</span>,</span>
<span>                    Exons <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">exons</span>,</span>
<span>                    Introns <span class="op">=</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">introns</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="sum-pips-within-annotation-categories">Sum PIPs within annotation categories<a class="anchor" aria-label="anchor" href="#sum-pips-within-annotation-categories"></a>
</h3>
<p>Unlike (below), it is OK to have overlapping annotations here. If a SNP is in multiple annotation categories, it will be assigned to the first ordered category.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum.pip.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partition_pip_annots.html">partition_pip_annots</a></span><span class="op">(</span><span class="va">finemapstats</span>, <span class="va">annots.list</span><span class="op">)</span></span></code></pre></div>
<p>Sum of PIPs in each annotation category:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum.pips</span> <span class="op">&lt;-</span> <span class="va">sum.pip.res</span><span class="op">$</span><span class="va">sum.pips</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#            OCRs         UTRs        Exons     Introns       others</span></span>
<span><span class="co"># 7  7.807975e-01 0.0000000000 0.000000e+00 0.219071200 0.0000000000</span></span>
<span><span class="co"># 15 9.831023e-01 0.0000000000 0.000000e+00 0.006548894 0.0102402696</span></span>
<span><span class="co"># 25 4.845244e-03 0.0006967073 1.186105e-05 0.992016899 0.0006346161</span></span>
<span><span class="co"># 31 3.931823e-05 0.0000000000 0.000000e+00 0.999293385 0.0000000000</span></span>
<span><span class="co"># 32 6.800833e-01 0.0000000000 0.000000e+00 0.141160792 0.1787247317</span></span>
<span><span class="co"># 33 5.780216e-01 0.0239950449 0.000000e+00 0.314905904 0.0830566374</span></span></code></pre>
<p>Obtain a matrix of the proportion of PIPs in each annotation category.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">locus.order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sum.pips</span>, <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">OCRs</span>, <span class="va">UTRs</span>, <span class="va">Exons</span>, <span class="va">Introns</span>, <span class="va">others</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sum.pips</span> <span class="op">&lt;-</span> <span class="va">sum.pips</span><span class="op">[</span><span class="va">locus.order</span>,<span class="op">]</span></span>
<span><span class="va">prop.pip.mat</span> <span class="op">&lt;-</span> <span class="va">sum.pips</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">prop.pip.mat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#           OCRs UTRs Exons      Introns       others</span></span>
<span><span class="co"># 728  1.0000000    0     0 0.000000e+00 0.000000e+00</span></span>
<span><span class="co"># 862  1.0000000    0     0 0.000000e+00 0.000000e+00</span></span>
<span><span class="co"># 1431 1.0000000    0     0 0.000000e+00 0.000000e+00</span></span>
<span><span class="co"># 660  1.0000000    0     0 0.000000e+00 0.000000e+00</span></span>
<span><span class="co"># 624  0.9999886    0     0 0.000000e+00 1.135265e-05</span></span>
<span><span class="co"># 1416 0.9999696    0     0 3.041734e-05 0.000000e+00</span></span></code></pre>
<p>We can make a structure plot to show the proportion of PIPs in each annotation category.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">categories</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OCRs"</span>, <span class="st">"UTRs"</span>, <span class="st">"Exons"</span>, <span class="st">"Introns"</span>, <span class="st">"others"</span><span class="op">)</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>OCRs <span class="op">=</span> <span class="st">"#E18727FF"</span>, UTRs <span class="op">=</span> <span class="st">"#238b45"</span>, Exons <span class="op">=</span>  <span class="st">"#bee6af"</span>, Introns <span class="op">=</span> <span class="st">"#B09C85FF"</span>, others <span class="op">=</span> <span class="st">"#aaaaaa"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pip_structure_plot.html">pip_structure_plot</a></span><span class="op">(</span><span class="va">prop.pip.mat</span>, categories <span class="op">=</span> <span class="va">categories</span>, colors <span class="op">=</span> <span class="va">colors</span><span class="op">)</span></span></code></pre></div>
<p><img src="partition_pip_tutorial_files/figure-html/structure-plot-annots-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="partition-pips-into-disjoint-ocrs-for-different-cell-types">Partition PIPs into disjoint OCRs for different cell types<a class="anchor" aria-label="anchor" href="#partition-pips-into-disjoint-ocrs-for-different-cell-types"></a>
</h3>
<p>We can partition PIPs into disjoint OCRs for different cell types.</p>
<p>Load a list of GRanges objects containing disjoint OCRs for different cell types.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">disjoint.OCRs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"disjoint.OCRs.hg19.grlist.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Sum PIPs within cell-type specific OCRs.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum.pip.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/partition_pip_regions.html">partition_pip_regions</a></span><span class="op">(</span><span class="va">finemapstats</span>, <span class="va">disjoint.OCRs</span><span class="op">)</span></span></code></pre></div>
<p>Sum of PIPs in each cell type OCR category:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sum.pips</span> <span class="op">&lt;-</span> <span class="va">sum.pip.res</span><span class="op">$</span><span class="va">sum.pips</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#    Cardiomyocyte  Endothelial Fibroblast Lymphoid      Myeloid Pericyte</span></span>
<span><span class="co"># 7   7.413873e-01 0.0000000000 0.00000000        0 0.000000e+00        0</span></span>
<span><span class="co"># 15  5.155213e-02 0.0002890132 0.01230289        0 0.000000e+00        0</span></span>
<span><span class="co"># 25  5.315708e-04 0.0000000000 0.00000000        0 1.364875e-05        0</span></span>
<span><span class="co"># 31  3.931823e-05 0.0000000000 0.00000000        0 0.000000e+00        0</span></span>
<span><span class="co"># 32  6.799067e-01 0.0000000000 0.00000000        0 0.000000e+00        0</span></span>
<span><span class="co"># 33  2.886927e-01 0.0000000000 0.00000000        0 0.000000e+00        0</span></span>
<span><span class="co">#      Shared 2-3    Shared 4+</span></span>
<span><span class="co"># 7  0.000000e+00 0.0000000000</span></span>
<span><span class="co"># 15 4.332407e-05 0.0093814102</span></span>
<span><span class="co"># 25 2.887397e-03 0.0002959630</span></span>
<span><span class="co"># 31 0.000000e+00 0.0000000000</span></span>
<span><span class="co"># 32 0.000000e+00 0.0001766484</span></span>
<span><span class="co"># 33 1.821428e-03 0.0006501397</span></span></code></pre>
<p>Select loci with total PIPs in OCR &gt; 0.25, and compute the proportion of PIPs partitioned in each cell type category.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># reorder the loci to match the previous figure</span></span>
<span><span class="va">sum.pips</span> <span class="op">&lt;-</span> <span class="va">sum.pips</span><span class="op">[</span><span class="va">locus.order</span>, <span class="op">]</span></span>
<span><span class="co"># filter loci with total PIPs in OCR &gt; 0.25</span></span>
<span><span class="va">sum.pips.filtered</span> <span class="op">&lt;-</span> <span class="va">sum.pips</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">sum.pips</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.25</span>,<span class="op">]</span></span>
<span><span class="va">prop.pip.mat</span> <span class="op">&lt;-</span> <span class="va">sum.pips.filtered</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">sum.pips.filtered</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">prop.pip.mat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#      Cardiomyocyte Endothelial Fibroblast Lymphoid      Myeloid Pericyte</span></span>
<span><span class="co"># 862      1.0000000           0          0        0 0.000000e+00        0</span></span>
<span><span class="co"># 1431     0.9999189           0          0        0 0.000000e+00        0</span></span>
<span><span class="co"># 660      0.0000000           0          0        0 0.000000e+00        0</span></span>
<span><span class="co"># 624      1.0000000           0          0        0 0.000000e+00        0</span></span>
<span><span class="co"># 1682     1.0000000           0          0        0 0.000000e+00        0</span></span>
<span><span class="co"># 645      0.9999885           0          0        0 1.153082e-05        0</span></span>
<span><span class="co">#      Shared 2-3    Shared 4+</span></span>
<span><span class="co"># 862           0 0.000000e+00</span></span>
<span><span class="co"># 1431          0 8.110793e-05</span></span>
<span><span class="co"># 660           0 1.000000e+00</span></span>
<span><span class="co"># 624           0 0.000000e+00</span></span>
<span><span class="co"># 1682          0 0.000000e+00</span></span>
<span><span class="co"># 645           0 0.000000e+00</span></span></code></pre>
<p>We can make a structure plot to show the proportion of PIPs in each cell type category.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">categories</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cardiomyocyte"</span>, <span class="st">"Endothelial"</span>, <span class="st">"Fibroblast"</span>, <span class="st">"Lymphoid"</span>, </span>
<span>                <span class="st">"Myeloid"</span>, <span class="st">"Pericyte"</span>, <span class="st">"Shared 2-3"</span>, <span class="st">"Shared 4+"</span><span class="op">)</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#b22222"</span>, <span class="st">"#8DD3C7"</span>, <span class="st">"#BEBADA"</span>, <span class="st">"#FB8072"</span>, </span>
<span>            <span class="st">"#80B1D3"</span>, <span class="st">"#B3DE69"</span>, <span class="st">"royalblue"</span>, <span class="st">"#003C86"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pip_structure_plot.html">pip_structure_plot</a></span><span class="op">(</span><span class="va">prop.pip.mat</span>, categories <span class="op">=</span> <span class="va">categories</span>, colors <span class="op">=</span> <span class="va">colors</span><span class="op">)</span></span></code></pre></div>
<p><img src="partition_pip_tutorial_files/figure-html/structure-plot-OCRs-1.png" width="960"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
