<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial for making track plots • mapgen</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for making track plots">
<meta property="og:description" content="mapgen">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="track_plot_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial for making track plots</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/vignettes/track_plot_tutorial.Rmd" class="external-link"><code>vignettes/track_plot_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>track_plot_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<ul>
<li>Fine-mapping summary statistics.</li>
<li>Gene annotations.</li>
<li>Functional annotation data, e.g.: chromatin loops, open chromatin regions, histone modification data, etc.</li>
</ul>
<p>Please have the following Bioconductor packages installed: <code>GenomicFeatures</code>, <code>rtracklayer</code>, <code>Gviz</code>, <code>GenomicInteractions</code>, <code>AnnotationDbi</code>, <code>org.Hs.eg.db</code>.</p>
<p>Load packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span> <span class="co">#  Making and manipulating annotations</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span> <span class="co"># Import annotation data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ivanek/Gviz" class="external-link">Gviz</a></span><span class="op">)</span> <span class="co"># R package used to visualize track plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ComputationalRegulatoryGenomicsICL/GenomicInteractions/" class="external-link">GenomicInteractions</a></span><span class="op">)</span> <span class="co"># visualize HiC loops</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/AnnotationDbi" class="external-link">AnnotationDbi</a></span><span class="op">)</span> <span class="co"># match gene ID to gene symbol</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span> <span class="co"># match gene ID to gene symbol</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/mapgen" class="external-link">mapgen</a></span><span class="op">)</span></span></code></pre></div>
<p>If you are in Xin He lab at U Chicago, you can access the example data on RCC:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trackdata.dir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/mapgen/example_data/trackplot"</span></span></code></pre></div>
<div class="section level3">
<h3 id="fine-mapping-summary-statistics">Fine-mapping summary statistics<a class="anchor" aria-label="anchor" href="#fine-mapping-summary-statistics"></a>
</h3>
<p>Here we use the fine-mapping summary statistics from the AFib study.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"AF.finemapping.sumstats.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">finemapstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_finemapping_sumstats.html">process_finemapping_sumstats</a></span><span class="op">(</span><span class="va">finemapstats</span>, </span>
<span>                                             snp <span class="op">=</span> <span class="st">'snp'</span>, </span>
<span>                                             chr <span class="op">=</span> <span class="st">'chr'</span>, </span>
<span>                                             pos <span class="op">=</span> <span class="st">'pos'</span>, </span>
<span>                                             pip <span class="op">=</span> <span class="st">'susie_pip'</span>, </span>
<span>                                             pval <span class="op">=</span> <span class="st">'pval'</span>, </span>
<span>                                             zscore <span class="op">=</span> <span class="st">'zscore'</span>, </span>
<span>                                             cs <span class="op">=</span> <span class="st">'cs'</span>, </span>
<span>                                             locus <span class="op">=</span> <span class="st">'locus'</span>,  </span>
<span>                                             pip.thresh <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Processing fine-mapping summary statistics ...</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="gene-annotations">Gene annotations<a class="anchor" aria-label="anchor" href="#gene-annotations"></a>
</h3>
<p>You can download gene annotations (GTF file) from <a href="https://www.gencodegenes.org" class="external-link">GENCODE</a>, and make gene annotations using the <code><a href="../reference/make_genomic_annots.html">make_genomic_annots()</a></code> function with the GTF file.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="st">'/project2/xinhe/shared_data/gencode/gencode.v19.annotation.gtf.gz'</span></span>
<span><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_genomic_annots.html">make_genomic_annots</a></span><span class="op">(</span><span class="va">gtf_file</span><span class="op">)</span></span></code></pre></div>
<p>We included gene annotations (hg19) in the package, downloaded from <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz" class="external-link">GENCODE release 19</a>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genomic.annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genomic.annots.hg19.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene.annots</span> <span class="op">&lt;-</span> <span class="va">genomic.annots</span><span class="op">$</span><span class="va">genes</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="txdb-database">TxDb database<a class="anchor" aria-label="anchor" href="#txdb-database"></a>
</h3>
<p>We can build a <code>TxDb</code> database (“.sqlite”) using the GTF file, and use  to load the <code>TxDb</code> database.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="st">'/project2/xinhe/shared_data/gencode/gencode.v19.annotation.gtf.gz'</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span><span class="va">gtf_file</span>, format <span class="op">=</span> <span class="st">"gtf"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">saveDb</a></span><span class="op">(</span><span class="va">txdb</span>, <span class="st">"gencode.v19.annotation.gtf.sqlite"</span><span class="op">)</span></span></code></pre></div>
<p>If you are in Xin He lab at UChicago, you can access the gene annotations and <code>TxDb</code> database from RCC.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="st">'/project2/xinhe/shared_data/gencode/gencode.v19.annotation.gtf.gz'</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/gencode/gencode.v19.annotation.gtf.sqlite"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genomic-annotations">Genomic annotations<a class="anchor" aria-label="anchor" href="#genomic-annotations"></a>
</h3>
<p>Load promoter-capture HiC (PC-HiC) data from cardiomyocytes (CMs).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcHiC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"pcHiC.CM.gr.rds"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pcHiC</span> <span class="op">&lt;-</span> <span class="va">pcHiC</span><span class="op">[</span><span class="va">pcHiC</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">gene.annots</span><span class="op">$</span><span class="va">gene_name</span>, <span class="op">]</span> <span class="co"># restrict to protein coding genes</span></span></code></pre></div>
<p>Load ABC scores from heart ventricle (from Nasser <em>et al.</em> <em>Nature</em> 2021).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"heart_ventricle-ENCODE_ABC.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"mapgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_ABC.html">process_ABC</a></span><span class="op">(</span><span class="va">ABC</span>, full.element <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ABC</span> <span class="op">&lt;-</span> <span class="va">ABC</span><span class="op">[</span><span class="va">ABC</span><span class="op">$</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">gene.annots</span><span class="op">$</span><span class="va">gene_name</span>, <span class="op">]</span> <span class="co"># restrict to protein coding genes</span></span>
<span><span class="va">ABC</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="va">ABC</span><span class="op">$</span><span class="va">score</span> <span class="op">*</span> <span class="fl">100</span> <span class="co"># scale to visualize the ABC scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ABC</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames        ranges strand | promoter_start promoter_end   gene_name</span></span>
<span><span class="co">##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;integer&gt;    &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 888243-888743      * |         894679       894679       NOC2L</span></span>
<span><span class="co">##   [2]     chr1 908361-908861      * |         895966       895966      KLHL17</span></span>
<span><span class="co">##   [3]     chr1 908361-908861      * |         901876       901876     PLEKHN1</span></span>
<span><span class="co">##           score</span></span>
<span><span class="co">##       &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]    1.5224</span></span>
<span><span class="co">##   [2]    1.7673</span></span>
<span><span class="co">##   [3]    4.1100</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Load H3K27ac and DHS <code>.bed</code> files.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">H3K27ac_peaks</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">trackdata.dir</span>, <span class="st">"H3K27ac.heart.concat.hg19.bed.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">DHS_peaks</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">trackdata.dir</span>, <span class="st">"FetalHeart_E083.DNase.hg19.narrowPeak.bed.gz"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Load ATAC-seq counts data (the data should be in <code>wig</code>, <code>bigWig/bw</code>, <code>bedGraph</code>, or <code>bam</code> format)</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CM_counts</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">trackdata.dir</span>, <span class="st">"Cardiomyocyte.atac.hg19.bedGraph.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Endo_counts</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">trackdata.dir</span>, <span class="st">"Endothelial.atac.hg19.bedGraph.gz"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">Fibro_counts</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">trackdata.dir</span>, <span class="st">"Fibroblast.atac.hg19.bedGraph.gz"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="ld-reference-panel">LD reference panel<a class="anchor" aria-label="anchor" href="#ld-reference-panel"></a>
</h3>
<p>If you want to visualize r^2 between SNPs, we need a reference panel in a <code>bigSNP</code> object. (we will implement an option that takes LD matrix as input)</p>
<p>If you don’t provide the <code>bigSNP</code> object, the SNPs in the GWAS track will be plotted in the same color.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span> <span class="co"># loading reference genotype for LD calculation</span></span></code></pre></div>
<p>If you are in the He lab at UChicago, you can load the <code>bigSNP</code> object from the 1000 Genomes (1KG) European population.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bigSNP</span> <span class="op">&lt;-</span> <span class="fu">bigsnpr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span>rdsfile <span class="op">=</span> <span class="st">'/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## FBM from an old version? Reconstructing..</span></span></code></pre>
<pre><code><span><span class="co">## You should use `snp_save()`.</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="make-track-plots">Make track plots<a class="anchor" aria-label="anchor" href="#make-track-plots"></a>
</h2>
<p>Plot HCN4 locus in the genomic region “chr15:73610000-73700000”</p>
<p>Highlight SNP “rs7172038”</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"CM"</span> <span class="op">=</span> <span class="va">CM_counts</span>, <span class="st">"Endo"</span> <span class="op">=</span> <span class="va">Endo_counts</span>, <span class="st">"Fibro"</span> <span class="op">=</span> <span class="va">Fibro_counts</span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"H3K27ac"</span> <span class="op">=</span> <span class="va">H3K27ac_peaks</span>, <span class="st">"DHS"</span> <span class="op">=</span> <span class="va">DHS_peaks</span><span class="op">)</span></span>
<span><span class="va">loops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"ABC"</span> <span class="op">=</span> <span class="va">ABC</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/track_plot.html">track_plot</a></span><span class="op">(</span><span class="va">finemapstats</span>,</span>
<span>           region <span class="op">=</span> <span class="st">"chr15:73610000-73700000"</span>,</span>
<span>           <span class="va">gene.annots</span>,</span>
<span>           bigSNP <span class="op">=</span> <span class="va">bigSNP</span>,</span>
<span>           txdb <span class="op">=</span> <span class="va">txdb</span>,</span>
<span>           counts <span class="op">=</span> <span class="va">counts</span>,</span>
<span>           peaks <span class="op">=</span> <span class="va">peaks</span>,</span>
<span>           loops <span class="op">=</span> <span class="va">loops</span>,</span>
<span>           genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>           filter_loop_genes <span class="op">=</span> <span class="st">"HCN4"</span>,</span>
<span>           highlight_snps <span class="op">=</span> <span class="st">"topSNP"</span>, </span>
<span>           counts.color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"purple"</span><span class="op">)</span>,</span>
<span>           peaks.color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"navy"</span>, <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>           loops.color <span class="op">=</span> <span class="st">"gray"</span>, </span>
<span>           genelabel.side <span class="op">=</span> <span class="st">"above"</span>,</span>
<span>           verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 463 snps included.</span></span>
<span><span class="co">## Color SNPs in PIP track by loci. </span></span>
<span><span class="co">## Adding CM track...</span></span>
<span><span class="co">## Adding Endo track...</span></span>
<span><span class="co">## Adding Fibro track...</span></span>
<span><span class="co">## Adding H3K27ac track...</span></span>
<span><span class="co">## Adding DHS track...</span></span>
<span><span class="co">## Adding ABC track...</span></span>
<span><span class="co">## Only show ABC loops linked to gene: HCN4 </span></span>
<span><span class="co">## Making gene track object using txdb database ...</span></span></code></pre>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">## Highlight SNP: rs7172038 </span></span>
<span><span class="co">## Making track plot ...</span></span></code></pre>
<p><img src="track_plot_tutorial_files/figure-html/make-HCN4-trackplot-1.png" width="1152"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
