<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ mapgen</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="LD_diagnosis_susie_rss.html">LD_diagnosis_susie_rss()</a></code> </p>
        </td>
        <td><p>Perform diagnosis to check the consistency between the GWAS z-scores and LD matrix,
and detect problematic z-scores (e.g. allele switch issue) using
diagnostic functions from SuSiE RSS.</p></td>
      </tr><tr><td>
          <p><code><a href="assign_snp_locus.html">assign_snp_locus()</a></code> </p>
        </td>
        <td><p>Assign GWAS SNPs to LD blocks</p></td>
      </tr><tr><td>
          <p><code><a href="block_view_summary.html">block_view_summary()</a></code> </p>
        </td>
        <td><p>LD block view summary table</p></td>
      </tr><tr><td>
          <p><code><a href="clean_sumstats.html">clean_sumstats()</a></code> </p>
        </td>
        <td><p>Cleans summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="compute_gene_pip.html">compute_gene_pip()</a></code> </p>
        </td>
        <td><p>Compute gene PIPs based on fine-mapping result and functional annotations.</p></td>
      </tr><tr><td>
          <p><code><a href="extract_gene_level_result.html">extract_gene_level_result()</a></code> </p>
        </td>
        <td><p>Extract gene-level result from SNP-level gene mapping result</p></td>
      </tr><tr><td>
          <p><code><a href="find_nearest_genes.html">find_nearest_genes()</a></code> </p>
        </td>
        <td><p>Find the nearest genes for top SNPs in each locus.</p></td>
      </tr><tr><td>
          <p><code><a href="gene_cs.html">gene_cs()</a></code> </p>
        </td>
        <td><p>Obtain credible gene sets from SNP-level gene mapping table</p></td>
      </tr><tr><td>
          <p><code><a href="gene_manhattan_plot.html">gene_manhattan_plot()</a></code> </p>
        </td>
        <td><p>Make gene Manhattan plot</p></td>
      </tr><tr><td>
          <p><code><a href="gene_view_summary.html">gene_view_summary()</a></code> </p>
        </td>
        <td><p>Gene view summary table</p></td>
      </tr><tr><td>
          <p><code><a href="get_UKBB_region_info.html">get_UKBB_region_info()</a></code> </p>
        </td>
        <td><p>Get region info with filenames of LD matrices and variant information</p></td>
      </tr><tr><td>
          <p><code><a href="get_locus_level_gene_pip.html">get_locus_level_gene_pip()</a></code> </p>
        </td>
        <td><p>Get locus level gene PIP</p></td>
      </tr><tr><td>
          <p><code><a href="harmonize_sumstats_LD.html">harmonize_sumstats_LD()</a></code> </p>
        </td>
        <td><p>Harmonize GWAS summary statistics with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="load_UKBB_LDREF.html">load_UKBB_LDREF()</a></code> </p>
        </td>
        <td><p>Load UK Biobank LD reference matrix and variant information</p></td>
      </tr><tr><td>
          <p><code><a href="make_genomic_annots.html">make_genomic_annots()</a></code> </p>
        </td>
        <td><p>Make genomic annotations from a GTF file</p></td>
      </tr><tr><td>
          <p><code><a href="match_gwas_LDREF.html">match_gwas_LDREF()</a></code> </p>
        </td>
        <td><p>Match GWAS sumstats with LD reference files. Only keep variants included in
LD reference.</p></td>
      </tr><tr><td>
          <p><code><a href="match_gwas_bigsnp.html">match_gwas_bigsnp()</a></code> </p>
        </td>
        <td><p>Match alleles between GWAS summary statistics and bigSNP reference panel.</p></td>
      </tr><tr><td>
          <p><code><a href="merge_susie_sumstats.html">merge_susie_sumstats()</a></code> </p>
        </td>
        <td><p>merges SuSiE results with original summary statistics data frame</p></td>
      </tr><tr><td>
          <p><code><a href="nearby_interactions.html">nearby_interactions()</a></code> </p>
        </td>
        <td><p>Get nearby interactions for enhancer regions near promoters</p></td>
      </tr><tr><td>
          <p><code><a href="partition_pip_annots.html">partition_pip_annots()</a></code> </p>
        </td>
        <td><p>Partition PIPs into functional annotation categories.</p></td>
      </tr><tr><td>
          <p><code><a href="partition_pip_regions.html">partition_pip_regions()</a></code> </p>
        </td>
        <td><p>Partition PIPs into disjoint functional annotation regions</p></td>
      </tr><tr><td>
          <p><code><a href="pip_structure_plot.html">pip_structure_plot()</a></code> </p>
        </td>
        <td><p>Make a structure plot of partitioned PIP by locus</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_susie_data_with_torus_result.html">prepare_susie_data_with_torus_result()</a></code> </p>
        </td>
        <td><p>Prepare summary statistics with TORUS SNP-level priors as
input data for SuSiE</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_torus_input_files.html">prepare_torus_input_files()</a></code> </p>
        </td>
        <td><p>Prepare TORUS input files</p></td>
      </tr><tr><td>
          <p><code><a href="process_ABC.html">process_ABC()</a></code> </p>
        </td>
        <td><p>Process ABC scores and save as a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="process_finemapping_sumstats.html">process_finemapping_sumstats()</a></code> </p>
        </td>
        <td><p>Process fine-mapping summary statistics data</p></td>
      </tr><tr><td>
          <p><code><a href="process_gwas_sumstats.html">process_gwas_sumstats()</a></code> </p>
        </td>
        <td><p>Process GWAS summary statistics and harmonize with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="process_loop_data.html">process_loop_data()</a></code> </p>
        </td>
        <td><p>Process chromatin loop data and save as a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="process_pcHiC.html">process_pcHiC()</a></code> </p>
        </td>
        <td><p>Process PC-HiC data and save as a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="read_LD.html">read_LD()</a></code> </p>
        </td>
        <td><p>read LD matrix data by file format</p></td>
      </tr><tr><td>
          <p><code><a href="read_LD_SNP_info.html">read_LD_SNP_info()</a></code> </p>
        </td>
        <td><p>Read all SNP info in LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="run_finemapping.html">run_finemapping()</a></code> </p>
        </td>
        <td><p>Run functional fine-mapping using summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="run_torus.html">run_torus()</a></code> </p>
        </td>
        <td><p>Run enrichment analysis and compute SNP-level priors using TORUS</p></td>
      </tr><tr><td>
          <p><code><a href="snp_view_summary.html">snp_view_summary()</a></code> </p>
        </td>
        <td><p>SNP view summary table</p></td>
      </tr><tr><td>
          <p><code><a href="susie_finemap_region.html">susie_finemap_region()</a></code> </p>
        </td>
        <td><p>Run fine-mapping for one region with SuSiE using summary statistics (susie_rss)</p></td>
      </tr><tr><td>
          <p><code><a href="track_plot.html">track_plot()</a></code> </p>
        </td>
        <td><p>Make gene track plot using Gviz</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

