<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Make gene track plot using Gviz — track_plot • mapgen</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Make gene track plot using Gviz — track_plot"><meta property="og:description" content="Make gene track plot using Gviz"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mapgen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.5.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data_preparation_tutorial.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/enrichment_tutorial.html">Enrichment analysis</a>
    </li>
    <li>
      <a href="../articles/finemapping_tutorial.html">Functional fine-mapping</a>
    </li>
    <li>
      <a href="../articles/finemapping_ukbb_ld_diagnosis.html">Fine-mapping and LD diagnosis with UK Biobank reference LD matrices</a>
    </li>
    <li>
      <a href="../articles/gene_mapping_tutorial.html">Gene mapping</a>
    </li>
    <li>
      <a href="../articles/partition_pip_tutorial.html">PIP partitioning by annotation categories</a>
    </li>
    <li>
      <a href="../articles/track_plot_tutorial.html">Making track plots</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/mapgen/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make gene track plot using Gviz</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/mapgen/blob/HEAD/R/plots.R" class="external-link"><code>R/plots.R</code></a></small>
    <div class="hidden name"><code>track_plot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Make gene track plot using Gviz</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">track_plot</span><span class="op">(</span></span>
<span>  <span class="va">finemapstats</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  <span class="va">gene.annots</span>,</span>
<span>  bigSNP <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  txdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  loops <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  filter_loop_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_loop_snps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  color_piptrack_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"locus"</span>, <span class="st">"cs"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  highlight_snps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  counts.ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  <span class="va">counts.color</span>,</span>
<span>  peaks.color <span class="op">=</span> <span class="st">"navy"</span>,</span>
<span>  loops.color <span class="op">=</span> <span class="st">"gray"</span>,</span>
<span>  highlight.color <span class="op">=</span> <span class="st">"pink"</span>,</span>
<span>  genelabel.side <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"above"</span>, <span class="st">"below"</span>, <span class="st">"left"</span>, <span class="st">"right"</span><span class="op">)</span>,</span>
<span>  track.sizes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rotation.title <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>finemapstats</dt>
<dd><p>A GRanges object of processed finemapping summary statistics</p></dd>


<dt>region</dt>
<dd><p>The genomic region to visualize in the format of "chr:start-end".</p></dd>


<dt>gene.annots</dt>
<dd><p>A GRanges object of gene annotations</p></dd>


<dt>bigSNP</dt>
<dd><p>A <code>bigsnpr</code> object attached via bigsnpr::snp_attach()</p></dd>


<dt>txdb</dt>
<dd><p>A <code>txdb</code> object of gene annotations</p></dd>


<dt>counts</dt>
<dd><p>A list of counts data to display as quantitativ data tracks</p></dd>


<dt>peaks</dt>
<dd><p>A list of peaks to display as binary data tracks</p></dd>


<dt>loops</dt>
<dd><p>A list of chromatin loops, e.g. PC-HiC, ABC, etc.</p></dd>


<dt>genome</dt>
<dd><p>Genome assembly version (default: "hg19").</p></dd>


<dt>filter_loop_genes</dt>
<dd><p>A vector of gene names. Only show loops connected to the genes.</p></dd>


<dt>filter_loop_snps</dt>
<dd><p>A vector of SNP IDs. Only show loops connected to the SNPs.</p></dd>


<dt>filter_protein_coding_genes</dt>
<dd><p>Logical. If TRUE, only shows protein coding gene.</p></dd>


<dt>color_piptrack_by</dt>
<dd><p>Color SNPs in the PIP track by
<code>locus</code>, <code>cs</code> (credible sets), or <code>none</code> (same color).</p></dd>


<dt>highlight_snps</dt>
<dd><p>SNPs (rsIDs) to highlight. Or highlight top SNP ('topSNP').</p></dd>


<dt>counts.ylim</dt>
<dd><p>ylim range (default: between 0 and 1) for the <code>counts</code> tracks</p></dd>


<dt>counts.color</dt>
<dd><p>Colors for the <code>counts</code> tracks</p></dd>


<dt>peaks.color</dt>
<dd><p>Colors for the <code>peak</code> tracks</p></dd>


<dt>highlight.color</dt>
<dd><p>Colors for the highlighted SNPs</p></dd>


<dt>genelabel.side</dt>
<dd><p>Side to put gene labels,
options are: 'above' (default), 'below', 'left', 'right'.</p></dd>


<dt>track.sizes</dt>
<dd><p>Sizes of the tracks</p></dd>


<dt>rotation.title</dt>
<dd><p>The rotation angle for the text in the track title</p></dd>


<dt>verbose</dt>
<dd><p>If TRUE, print detail messages for plotting</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kaixuan Luo, Alan Selewa, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

